kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/format/KafkaEntryFormatter.java

```diff
@@ -61,15 +65,40 @@ public DecodeResult decode(List<Entry> entries, byte magic) {
 
         // reset header information
         final List<ByteBuf> orderedByteBuf = new ArrayList<>();
+
         for (Entry entry : entries) {
             try {
                 long startOffset = MessageIdUtils.peekBaseOffsetFromEntry(entry);
                 final ByteBuf byteBuf = entry.getDataBuffer();
                 final MessageMetadata metadata = Commands.parseMessageMetadata(byteBuf);
                 if (isKafkaEntryFormat(metadata)) {
+                    byte batchMagic = byteBuf.getByte(byteBuf.readerIndex() + MAGIC_OFFSET);
                     byteBuf.setLong(byteBuf.readerIndex() + OFFSET_OFFSET, startOffset);
-                    byteBuf.setByte(byteBuf.readerIndex() + MAGIC_OFFSET, magic);
-                    orderedByteBuf.add(byteBuf.slice(byteBuf.readerIndex(), byteBuf.readableBytes()));
+                    ConvertedRecords<MemoryRecords> convertedRecords = null;
+
+                    // batch magic greater than the magic corresponding to the version requested by the client
+                    // need down converted
+                    if (batchMagic > magic) {
+                        MemoryRecords memoryRecords = MemoryRecords.readableRecords(ByteBufUtils.getNioBuffer(byteBuf));
+                        //down converted, batch magic will be set to client magic
+                        convertedRecords = memoryRecords.downConvert(magic, startOffset, time);
+                        log.trace("[{}:{}] downConvert record, start offset {}, entry magic: {}, client magic: {}"
+                                , entry.getLedgerId(), entry.getEntryId(), startOffset, batchMagic, magic);
+                    }
+                    if (convertedRecords != null) {
```

Comment:
Should we remove this check and move the block of `convertedRecords != null` into `batchMagic > magic` block?

Reply:
@BewareMyPower Thanks for your review, You are right, I have addressed your comment, PTAL